Added reviews for sessions

SaptakS opened the pull request:
I have added the database model, views and helper python functions so that it may function in the following way.

```
POST - "/sessions/:id/reviews" 
body - { 
"email" : "me@example.org",
"rating": 4.0, // can be from 1.0 to 5.0
"comment": "Lorem Ipsum"
}

RESPONSE:
200 OK

GET - "/sessions/:id/reviews"
Response - "
{
"reviews": [{
          "email" : "me@example.org",
          "rating": 4.0, 
         "comment": "Lorem Ipsum"
         },
        // More reviews
        ],
"average": 3.5
"total": 100
// More paging stuff
}
"
```

creativepsyco commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
+        return "200 OK\n"
+    else:
+        print "Error: Email and Rating are required\n"
+        return "Error: Email and Rating are required\n"
```
you should be returning a valid json and not just ok, plus it doesn't have the right status. See [JSONAPI](http://jsonapi.org/format/) to figure out the format. usually it should be of the form:

```json
{
  "errors": [
    {
      "status": "403",
      "source": { "pointer": "/data/attributes/secret-powers" },
      "detail": "Editing secret powers is not authorized on Sundays."
    },
    {
      "status": "422",
      "source": { "pointer": "/data/attributes/volume" },
      "detail": "Volume does not, in fact, go to 11."
    },
    {
      "status": "500",
      "source": { "pointer": "/data/attributes/reputation" },
      "title": "The backend responded with an error",
      "detail": "Reputation service not responding after three requests."
    }
  ]
}
```

SaptakS replied:
I didn't understand the source part of the json.. Can you please elaborate? I am working on the rest.

SaptakS replied:
For errors, should I use the `@app.errorhandler` of flask? In that case, it will send this error response in case of error in all the others APIs mentioned before. Is that ok?

creativepsyco commented:
A few pointers:
* The api should be hosted at: `api/v1/event/:event_id/sessions/:session_id/reviews`, it is no reason to move away from this format
* Pagination stuff is missing, I would recommend, giving 50 reviews and then page the rest. Look for pagination format here: http://jsonapi.org/examples/#pagination
* Data returned is not valid json and not contains the HTTP status code information.
* Migration file is missing. Please use `python manage.py db migrate` to generate a migration file for model changes. See [Readme](https://github.com/fossasia/open-event-orga-server/blob/master/README.md)

creativepsyco commented:
Also don't use `print` use the logger.info/warn/debug/verbose options.

creativepsyco commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
+    print reviews.count()
+
+    if reviews.count() > 0:
+        DataManager.update_review(data, reviews[0], session_id)
```
This seems wrong, you are supposed to change the review for a particular session, but you are seemingly updating the first review in your query after filtering them by email.

A session can have many reviews.
A review must belong to user and to a session.
A user can have many reviews for many sessions.

SaptakS replied:
Yeah.. I got that wrong.. I have to check for both the user and the session. Will make the change asap.

creativepsyco commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
+    """Post reviews for sessions"""
+    data = json.loads(request.data)
+    if "comment" not in data:
+            data["comment"] = ""
```
too much indentation here.

creativepsyco commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
+    if reviews.count() > 0:
+        DataManager.update_review(data, reviews[0], session_id)
+        return "200 OK\n"
+    elif "email" in data and "rating" in data:  
```
use proper validation procedures. including validity of emails.

SaptakS replied:
Won't that already be done while authentication from the client-side?

creativepsyco replied:
see validators.py, atleast have some pre-conditions, can't depend on the client to have all the checks.

SaptakS replied:
Ok.

creativepsyco commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
+    if "comment" not in data:
+            data["comment"] = ""
+
+    reviews = DataGetter.get_reviews_by_email(data["email"])
```
`get_reviews_by_email_and_session_id`

creativepsyco commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
 	cal.add_component(track)
 	return cal.to_ical()
 
+@app.route('/api/v1/session/<int:session_id>/reviews', methods=['GET'])
```
`PUT` method for updating existing review?

SaptakS replied:
The POST method will handle updating existing review.. Should I use PUT for update?

creativepsyco commented:
I am also thinking there needs to be some sort of a secret shared key between the server & client to prevent abuse of the API. Let me think about that.

leto commented:
@creativepsyco yes, very much agreed. API write operations need some kind of API key.  See #187 for my brain dump.

I would say that this PR can get merged before we have API keys figured out but we definitely need API keys before running anything in production or we are just asking for trouble. Once we have API keys implemented on the backend it should be a simple matter of adding a decorator to any API methods that do write operations.

SaptakS commented:
@creativepsyco The api is hosted in `api/v1/sessions/:session_id/reviews` because every session has a unique session_id. So is it needed to be like this: `api/v1/event/:event_id/sessions/:session_id/reviews`?

SaptakS commented:
Also, which data returned is not a valid JSON? The status or the GET data? For the GET data, I have used the format which has been used in the other APIs.

creativepsyco commented:
return `return "200 OK\n"`

and `"Error: Email and Rating are required\n"` is not valid json, you are returning this from the views.py file, it has to be a json string.

creativepsyco commented:
Plus has to be the correct status code, even when you are raising errors. 400 for the case of invalid data.

SaptakS commented on open_event/views/views.py:
```diff
@@ -278,6 +281,38 @@ def generate_icalender_track(track_id):
+        return "200 OK\n"
+    elif "email" in data and "rating" in data:  
+        DataManager.create_review(data, session_id)        
+        return "200 OK\n"
```
Just to clarify, do you mean my response should be something like:
```
return jsonify({
    "status" : 200,
    "detail" : "Review Added Successfully"
})
```

creativepsyco replied:
yes

On Thu, Feb 18, 2016 at 5:54 PM Saptak Sengupta <notifications@github.com>
wrote:

> In open_event/views/views.py
> <https://github.com/fossasia/open-event-orga-server/pull/186#discussion_r53292332>
> :
>
> > +@cross_origin()
> > +def post_reviews(session_id):
> > +    """Post reviews for sessions"""
> > +    data = json.loads(request.data)
> > +    if "comment" not in data:
> > +            data["comment"] = ""
> > +
> > +    reviews = DataGetter.get_reviews_by_email(data["email"])
> > +    print reviews.count()
> > +
> > +    if reviews.count() > 0:
> > +        DataManager.update_review(
>
> data, reviews[0], session_id)
> > +        return "200 OK\n"
> > +    elif "email" in data and "rating" in data:
> > +        DataManager.create_review(data, session_id)
> > +        return "200 OK\n"
>
> Just to clarify, do you mean my response should be something like:
>
> return jsonify({
>     "status" : 200,
>     "detail" : "Review Added Successfully"
> })
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fossasia/open-event-orga-server/pull/186/files#r53292332>
> .
>

SaptakS commented:
I have jsonified the status returns, added email validation and added retrival of rating from database using email and session-id so that a rating with a particular email and session-id is unique.

SaptakS commented:
Also, can you please clarify if I need to add a `PUT` request function as well because I have already implemented the updating functionality using `POST`. If needed please let me know.

SaptakS commented:
@leto @rafalkowalski can you please review this PR?

leto commented:
@SaptakS I need a migration file for the schema, the current PR by itself does not work:

`curl -v http://127.0.0.1:8001/api/v1/session/1/reviews` generates a 500 error:
`ProgrammingError: (psycopg2.ProgrammingError) relation "session_reviews" does not exist`

SaptakS commented:
@leto I am not sure what you want me to do. The migration file is already there. I have just changed the views so migration file will still be there I guess. Please let me know if I got it wrong.

leto commented:
What is the name of the migration file? I am talking about running the commands in the "Model updates" section of the README.md, in particular: ` python manage.py db migrate`

which should create a file in `migrations/versions/`. Make sure to `git add` it and make it part of the PR

SaptakS commented:
Ok. I will check it.

SaptakS commented:
Closing this PR due to some conflicting code.
